src/stdlib/object_from_array.rs

```diff
@@ -0,0 +1,130 @@
+use crate::compiler::prelude::*;
+
+fn make_object(values: Vec<Value>) -> Resolved {
+    values
+        .into_iter()
+        .map(make_key_value)
+        .collect::<Result<_, _>>()
+        .map(Value::Object)
+}
+
+fn make_key_value(value: Value) -> Result<(KeyString, Value), ExpressionError> {
+    value.try_array().map_err(Into::into).and_then(|array| {
+        let mut iter = array.into_iter();
+        let key: KeyString = match iter.next() {
+            None => return Err("array value too short".into()),
+            Some(Value::Bytes(key)) => String::from_utf8_lossy(&key).into(),
+            Some(_) => return Err("object keys must be strings".into()),
+        };
+        let value = iter.next().unwrap_or(Value::Null);
+        Ok((key, value))
+    })
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct ObjectFromArray;
+
+impl Function for ObjectFromArray {
+    fn identifier(&self) -> &'static str {
+        "object_from_array"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "values",
+            kind: kind::ARRAY,
+            required: true,
+        }]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "create an object from an array of keys/value pairs",
+            source: r#"object_from_array([["a", 1], ["b"], ["c", true, 3, 4]])"#,
+            result: Ok(r#"{"a": 1, "b": null, "c": true}"#),
+        }]
+    }
+
+    fn compile(
+        &self,
+        state: &TypeState,
+        _ctx: &mut FunctionCompileContext,
+        arguments: ArgumentList,
+    ) -> Compiled {
+        let values = ConstOrExpr::new(arguments.required("values"), state);
+
+        Ok(OFAFn { values }.as_expr())
+    }
+}
+
+#[derive(Clone, Debug)]
+struct OFAFn {
+    values: ConstOrExpr,
+}
+
+impl FunctionExpression for OFAFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        make_object(self.values.resolve(ctx)?.try_array()?)
+    }
+
+    fn type_def(&self, _state: &TypeState) -> TypeDef {
+        TypeDef::object(Collection::any())
+    }
+}
+
+#[derive(Clone, Debug)]
+enum ConstOrExpr {
```

Comment:
This is a straight-up clone of the construct in `zip`. Where would it make sense to move it into common code?

Reply:
Yes, we can introduce a `src/stdlib/utils.rs` (for the lack of a better name)